### PR TITLE
Extended Extend Retry HTTP Codes

### LINKED ISSUE #626: Logstash loses records on 500 status code
One of the ES nodes in our cluster crashed today, and while the cluster kept functioning some of the bulk insert requests performed by logstash failed with the below error:

[2017-07-19T04:47:46,560][WARN ][logstash.outputs.elasticsearch] Failed action. {:status=>500, :action=>["index", {:id=>"19e9c5dc-f23e-4adc-ad46-176869872768", :index=>"computing.network_latency.08f694be-0619-485c-a3a4-0d7dd25cf1ef-2017.29", :type=>"doc", :routing=>nil}, 2017-07-19T04:45:42.038Z %{host} %{message}], :response=>{"index"=>{"index"=>"computing.networklatency.08f694be-0619-485c-a3a4-0d7dd25cf1ef-2017.29", "type"=>"doc", "id"=>"19e9c5dc-f23e-4adc-ad46-176869872768", "status"=>500, "error"=>{"type"=>"node_not_connected_exception", "reason"=>"[data1-iil-005][10.184.95.5:9300] Node not connected"}}}}

This is obviously a temporary issue, so I would expect Logstash to retry - but it didn't, since 500 isn't a retryable code as far as logstash is concerned. Either way, this data was lost. We use Logstash 5.4, but even if we were using the DLQ feature (which is not clear to me in which version it is available), it wouldn't have helped since the elasticsearch output DLQ's only 400 and 404 errors.

Is this something that should be fixed in Logstash, or in ElasticSearch? While Logstash is correct to assume 500 means a permanent failure, this particular exception is somewhat of a gray area.

### PR DESCRIPTION
**Pull request by andrewvc**
This adds a commit on top of #628 . 

UPDATE:
This PR now retries ALL non-200 response codes indefinitely. 

Fixes #626

### CODE REVIEW

File: docs/index.asciidoc [hunk 1]
```diff
@@ -55,17 +55,26 @@ the new template is installed.
...
+- 429 (Too many requests) Graceful backoff indication by Elastic
+- 500 (Internal Server Error) when an ES node disconnect occurs this is temporarily returned
+- 502 (Bad gateway) During shutdown a proxy might return this
+- 503 (Service unavailable) Proxy returns an unavailable service
+- 504 (Gateway timed out) Proxy backend timed out
 
-NOTE: 409 exceptions are no longer retried. Please set a higher `retry_on_conflict` value if you experience 409 exceptions.
+The following document errors are handled as follows:
+- 400 and 404 errors are sent to the DLQ (if enabled)
+- 409 errors (conflict) are logged as a warn and dropped. If you'd like these retried
```
Comment (jordansissel) [inline docs/index.asciidoc]: incomplete sentence "If you'd like these retried <...missing...>"

File: docs/index.asciidoc [hunk 2]
```diff
@@ -55,17 +55,26 @@ the new template is installed.
...
-The following errors are retried infinitely:
+The following HTTP request errors to the bulk API are retried infinitely:
 
-- Network errors (inability to connect)
-- 429 (Too many requests) and
-- 503 (Service unavailable) errors
+- Network errors (Inability to connect)
+- 404 (Not found) Indicating an upstream proxy has no matching route
+- 429 (Too many requests) Graceful backoff indication by Elastic
+- 500 (Internal Server Error) when an ES node disconnect occurs this is temporarily returned
```
Comment (untergeek) [inline docs/index.asciidoc]: There should be a comma after "occurs"

File: docs/index.asciidoc [hunk 3]
```diff
@@ -55,17 +55,26 @@ the new template is installed.
...
+batch as a whole and for individual documents
 
-The following errors are retried infinitely:
+The following HTTP request errors to the bulk API are retried infinitely:
 
-- Network errors (inability to connect)
-- 429 (Too many requests) and
-- 503 (Service unavailable) errors
+- Network errors (Inability to connect)
+- 404 (Not found) Indicating an upstream proxy has no matching route
```
Comment (untergeek) [inline docs/index.asciidoc]: "Indicates an upstream..."

File: docs/index.asciidoc [hunk 4]
```diff
@@ -55,17 +55,26 @@ the new template is installed.
...
+- 429 (Too many requests) Graceful backoff indication by Elastic
```
Comment (untergeek) [inline docs/index.asciidoc]: "...from Elasticsearch"

File: docs/index.asciidoc [hunk 5]
```diff
@@ -55,17 +55,26 @@ the new template is installed.
...
+- 500 (Internal Server Error) when an ES node disconnect occurs this is temporarily returned
+- 502 (Bad gateway) During shutdown a proxy might return this
```
Comment (untergeek) [inline docs/index.asciidoc]: Comma after "shutdown"

File: docs/index.asciidoc (hunk 1 above)
Comment (untergeek) [inline docs/index.asciidoc]: Also, "logged as a warning"

**Comment by untergeek** [comment]
@andrewvc I like that you've differentiated between DOC, DLQ, and REQUEST codes.  Much easier to understand when reading the code.

File: docs/index.asciidoc (hunk 1 above)
Comment (andrewvc) [inline docs/index.asciidoc]: @jordansissel apparently my brain drops sentences after a certain point, I need to work on its retry logic too! :)

File: docs/index.asciidoc (hunk 2 above)
Comment (andrewvc) [inline docs/index.asciidoc]: Why? Are you sure? I don't think that's grammatical.

**Comment by jordansissel** [comment]
@andrewvc I'd like this to be backported to the 7.x branch also (so we can include with LS 5.6)

**Comment by andrewvc** [comment]
@untergeek / @jordansissel this is ready for another pass.

**Comment by andrewvc** [comment]
The more I think about this, the more I think we should retry *all* non-2xx responses from the `/_bulk/` endpoint. It should always return a 200.

**Comment by andrewvc** [comment]
I've updated this PR to treat all non-200 response codes as errors.

File: docs/index.asciidoc [hunk 6]
```diff
@@ -55,17 +55,18 @@ the new template is installed.
 
 ==== Retry Policy
 
-The retry policy has changed significantly in the 2.2.0 release.
+The retry policy has changed significantly in the 8.0.3 release.
```
Comment (jsvd) [inline docs/index.asciidoc]: s/8.0.3/8.1.1

File: docs/index.asciidoc [hunk 7]
```diff
@@ -55,17 +55,18 @@ the new template is installed.
...
+request are handled differently than error codes for individual documents.
 
-The following errors are retried infinitely:
+HTTP requests to the bulk API are expected to return a 200 response code. All other response codes are retried indefinitely.
 
-- Network errors (inability to connect)
-- 429 (Too many requests) and
-- 503 (Service unavailable) errors
+The following document errors are handled as follows:
+- 400 and 404 errors are sent to the DLQ (if enabled)
```
Comment (jsvd) [inline docs/index.asciidoc]: and what if DLQ is not enabled?

**Comment by andrewvc** [comment]
@jsvd I just pushed fixes for those. Am I good to merge? (I'm not sure if that was a final LGTM)

**Pull request closed without merging**